tls/s2n_server_hello.c

```diff
@@ -105,6 +105,12 @@ static int s2n_server_hello_parse(struct s2n_connection *conn)
     GUARD(s2n_stuffer_read_bytes(in, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
     GUARD(s2n_stuffer_read_bytes(in, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
 
+    /* If the client receives a second HelloRetryRequest in the same connection, it MUST send an error. */
+    if (s2n_hello_retry_validate(conn) == S2N_SUCCESS) {
+        ENSURE_POSIX(!s2n_is_hello_retry_handshake(conn), S2N_ERR_INVALID_HELLO_RETRY);
+        GUARD(s2n_set_hello_retry_required(conn));
```

Comment:
Why does this logic live here instead of in s2n_server_hello_retry_recv?

We should only need to detect that this server_hello is actually a hello_retry_request once. We're currently detecting and acting on it twice, once in this parse method and once in the recv method. Let's try to keep the HRR logic contained as much as possible.

Reply:
This logic lives here because while processing the keyshare extension in this parse method, the logic differs if the server hello receive is a hello retry message.

Reply:
Sounds reasonable, although we really shouldn't keep special-casing HRR logic within server_hello. We really need a single point where we switch over :\

In that case, just change the second s2n_hello_retry_validate to a is_hello_retry_message